dot/state/grandpa.go

```diff
@@ -26,40 +36,276 @@ var (
-func NewGrandpaStateFromGenesis(db chaindb.Database, genesisAuthorities []types.GrandpaVoter) (*GrandpaState, error) {
+func NewGrandpaStateFromGenesis(db chaindb.Database, bs *BlockState,
+	genesisAuthorities []types.GrandpaVoter) (*GrandpaState, error) {
 	grandpaDB := chaindb.NewTable(db, grandpaPrefix)
 	s := &GrandpaState{
-		db: grandpaDB,
+		db:                   grandpaDB,
+		blockState:           bs,
+		scheduledChangeRoots: new(changeTree),
+		forcedChanges:        new(orderedPendingChanges),
 	}
 
 	if err := s.setCurrentSetID(genesisSetID); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set current set id: %w", err)
 	}
 
 	if err := s.SetLatestRound(0); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set latest round: %w", err)
 	}
 
 	if err := s.setAuthorities(genesisSetID, genesisAuthorities); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set authorities: %w", err)
 	}
 
-	if err := s.setSetIDChangeAtBlock(genesisSetID, 0); err != nil {
-		return nil, err
+	if err := s.setChangeSetIDAtBlock(genesisSetID, 0); err != nil {
+		return nil, fmt.Errorf("cannot set change set id at block 0: %w", err)
 	}
 
 	return s, nil
 }
 
 // NewGrandpaState returns a new GrandpaState
-func NewGrandpaState(db chaindb.Database) (*GrandpaState, error) {
+func NewGrandpaState(db chaindb.Database, bs *BlockState) *GrandpaState {
 	return &GrandpaState{
-		db: chaindb.NewTable(db, grandpaPrefix),
-	}, nil
+		db:                   chaindb.NewTable(db, grandpaPrefix),
+		blockState:           bs,
+		scheduledChangeRoots: new(changeTree),
+		forcedChanges:        new(orderedPendingChanges),
+	}
+}
+
+// HandleGRANDPADigest receives a decoded GRANDPA digest and calls the right function to handles the digest
+func (s *GrandpaState) HandleGRANDPADigest(header *types.Header, digest scale.VaryingDataType) error {
+	switch val := digest.Value().(type) {
+	case types.GrandpaScheduledChange:
+		return s.addScheduledChange(header, val)
+	case types.GrandpaForcedChange:
+		return s.addForcedChange(header, val)
+	case types.GrandpaOnDisabled:
+		return nil
+	case types.GrandpaPause:
+		logger.Warn("GRANDPA Pause consensus message not implemented yet")
+		return nil
+	case types.GrandpaResume:
+		logger.Warn("GRANDPA Resume consensus message not implemented yet")
+		return nil
+	default:
+		return fmt.Errorf("not supported digest")
+	}
+}
+
+func (s *GrandpaState) addForcedChange(header *types.Header, fc types.GrandpaForcedChange) error {
+	auths, err := types.GrandpaAuthoritiesRawToAuthorities(fc.Auths)
+	if err != nil {
+		return fmt.Errorf("cannot parse GRANDPA authorities to raw authorities: %w", err)
+	}
+
+	pendingChange := &pendingChange{
+		bestFinalizedNumber: fc.BestFinalizedBlock,
+		nextAuthorities:     auths,
+		announcingHeader:    header,
+		delay:               fc.Delay,
+	}
+
+	err = s.forcedChanges.importChange(pendingChange, s.blockState.IsDescendantOf)
```

Comment:
seems really odd to always call `importChange, pruneChanges, etc..` with this `IsDescendantOf` function.  I think it makes more sense to make `orderedPendingChanges` a struct with a field of type `isDescendantOfFunc`.  That way you include it as state for the type `orderedPendingChanges`.

Reply:
I don't think keeping the function as a field in the `orderedPendingChanges` is good as we will keep this reference in memory until we garbage collect, but passing it as an argument then once the function returns the argument will be dropped

Reply:
I don't think there's much more overhead keeping a single reference to `s.blockState.IsDescendantOf` in a struct that includes an attribute for `[]pendingChange`.  The question you should ask yourself is, is the type `orderedPendingChanges`  intended to be stateful.  I would argue yes, since it requires a subset of functions of `blockState`.  You could just add it to a constructor for `orderedPendingChanges` when it's called in `NewGrandpaState`.